### PR TITLE
fix(diagnostics): prevent jumping signs column in insert mode

### LINKED ISSUE #26078: diagnostic signs disappear when entering insert mode
### Problem

LSP diagnostic signs disappear one second after entering insert mode, only to reappear after leaving insert mode.

This leads thel code "jumping" when insert mode  is entered due to the sign column disappearing. 

This is the closest thing I could find in the help (the default seems to be what I want, but the behavior is unwanted since diagnostics *are* updated):

    :helpgrep update_in_insert
    update_in_insert: (default false) Update diagnostics in
    Insert mode (if false, diagnostics are updated on
    InsertLeave)

Similar issues:

1. [https://www.reddit.com/r/neovim/comments/10z4q16/lsp\_diagnostics\_disappear\_completely\_in\_insert/](https://www.reddit.com/r/neovim/comments/10z4q16/lsp_diagnostics_disappear_completely_in_insert/)
2. [https://vi.stackexchange.com/questions/42126/neovim-how-to-keep-the-diagnostic-message-in-insert-mode](https://vi.stackexchange.com/questions/42126/neovim-how-to-keep-the-diagnostic-message-in-insert-mode)

### Expected behavior

- No diagnostics should disappear when entering insert mode.
  - (All diagnostics that were shown before entering insert mode should stay visible until insert mode is leaved)
- No new diagnostics should appear while in insert mode.

### PR DESCRIPTION
**Pull request by Werkov**
I'm using very basic and default LSP setup with pyright, i.e. update_in_insert=false (which turns out to also be my preferred config). The signs column disappearing a moment after entering insert mode is super annoying.
Thanks to @liskin's analysis in [1], I was lead to the problematic `.hide()` invocation. Simply disabling it preserves the signs column in insert mode but the signs would not be cleared when problematic lines are editted with commands (e.g. `dd` to remove the offending line would keep the stale mark).
After some trial and errors, I ended up with guarding the `.hide()` invocation only to normal mode which results in desired behavior for me and it should also maintain update_in_insert=true working like before.

Fix: #26078

[1] https://github.com/neovim/neovim/issues/26078

## Problem

#26078

TL;DR The signs column disappearing a moment after entering insert mode is super annoying.

## Solution

Do not hide signs in non-normal mode and preserve update_in_insert=true functionality.

### CODE REVIEW

**Comment by justinmk** [comment]
What is a one-sentence description of the problem?

**Comment by Werkov** [comment]
Re: problem description -- hopefully fixed. ~Waiting for tests to pass.~

File: runtime/lua/vim/diagnostic.lua [hunk 1]
```diff
@@ -2377,20 +2377,21 @@ function M.show(namespace, bufnr, diagnostics, opts)
     return
   end
 
-  M.hide(namespace, bufnr)
+  local opts_res = get_resolved_options(opts, namespace, bufnr)
+  local mode = api.nvim_get_mode()
+  if opts_res.update_in_insert or mode.mode:sub(1, 1) == 'n' then
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:2382]: This should probably be `!= 'i'` instead of `== 'n'`


```suggestion
  if opts_res.update_in_insert or mode.mode:sub(1, 1) ~= 'i' then
```

**Comment by gpanders** [comment]
Please also add a test case for this. The diagnostics code is already very complex and full of edge cases, test coverage helps ensure we don't regress.

**Comment by justinmk** [comment]
After discussion we are aligned on this approach: https://github.com/neovim/neovim/issues/26078#issuecomment-4102886772

**Review by themavik — COMMENTED** [review]
curl test neovim 38313 only xyz123

**Review by themavik — COMMENTED** [review]
trace test only

**Review by themavik — COMMENTED** [review]
vim.diagnostic.show: get_resolved_options has to run before the hide branch for insert mode. Skipping M.hide when update_in_insert is false fixes the jumping signs column; stale extmarks could linger if diagnostics clear mid-insert while scheduling is still pending.

**Comment by clason** [comment]
@themavik Please stop spamming unrequested "review" comments across PRs. You _will_ be blocked for automated LLM-generated comments; there will be no further warning.

**Comment by justinmk** [comment]
closing in favor of https://github.com/neovim/neovim/pull/38441
